services/cd-service/pkg/repository/transformer.go

```diff
@@ -827,126 +827,184 @@ func (c *DeployApplicationVersion) Transform(ctx context.Context, state *State)
+	return envGroupConfigs
+}
+
+func generateReleaseTrainResponse(envDeployedMsg, envSkippedMsg map[string]string, targetGroupName string) string {
+	resp := fmt.Sprintf("Release Train commit message for environment/environment group '%s':\n\n", targetGroupName)
+
+	// this to sort the env, to make sure that for the same input we always got the same output
+	envGroups := make([]string, 0, len(envDeployedMsg))
+	for env := range envDeployedMsg {
+		envGroups = append(envGroups, env)
 	}
+	sort.Strings(envGroups)
 
-	if !upstreamLatest {
-		_, ok = configs[upstreamEnvName]
-		if !ok {
-			return fmt.Sprintf("Could not find environment config for upstream env %q. Target env was %q", upstreamEnvName, targetEnvName), err
+	for _, env := range envGroups {
+		msg := envDeployedMsg[env]
+		resp += fmt.Sprintf("Release Train for '%s' environment:\n\n", env)
+		resp += msg
+		if skippedMsg, ok := envSkippedMsg[env]; ok {
+			resp += "Skipped services:\n"
+			resp += skippedMsg
 		}
 	}
+	return resp
+}
+
+func (c *ReleaseTrain) Transform(ctx context.Context, state *State) (string, error) {
+	var targetGroupName = c.Target
 
-	envLocks, err := state.GetEnvironmentLocks(targetEnvName)
+	configs, err := state.GetEnvironmentConfigs()
 	if err != nil {
-		return "", httperrors.InternalError(ctx, fmt.Errorf("could not get lock for environment %q: %w", targetEnvName, err))
+		return "", httperrors.InternalError(ctx, err)
 	}
-	if len(envLocks) > 0 {
-		return fmt.Sprintf("Target Environment '%s' is locked - exiting.", targetEnvName), nil
+	var envGroupConfigs = getEnvironmentGroupsEnvironmentsOrEnvironment(configs, targetGroupName)
+
+	if len(envGroupConfigs) == 0 {
+		return "", httperrors.PublicError(ctx, fmt.Errorf("could not find environment group or environment configs for '%v'", targetGroupName))
 	}
 
-	var apps []string
-	if upstreamLatest {
-		apps, err = state.GetApplications()
-		if err != nil {
-			return "", httperrors.InternalError(ctx, fmt.Errorf("could not get all applications for %q: %w", source, err))
+	// this to sort the env, to make sure that for the same input we always got the same output
+	envGroups := make([]string, 0, len(envGroupConfigs))
+	for env := range envGroupConfigs {
+		envGroups = append(envGroups, env)
+	}
+	sort.Strings(envGroups)
+
+	envDeployedMsg := make(map[string]string)
+	envSkippedMsg := make(map[string]string)
+	for _, envName := range envGroups {
+		envConfig := envGroupConfigs[envName]
+		if envConfig.Upstream == nil {
+			return fmt.Sprintf("Environment %q does not have upstream configured - exiting.", envName), nil
 		}
-	} else {
-		apps, err = state.GetEnvironmentApplications(upstreamEnvName)
-		if err != nil {
-			return "", httperrors.PublicError(ctx, fmt.Errorf("upstream environment (%q) does not have applications: %w", upstreamEnvName, err))
+		var upstreamLatest = envConfig.Upstream.Latest
+		var upstreamEnvName = envConfig.Upstream.Environment
+
+		if !upstreamLatest && upstreamEnvName == "" {
```

Comment:
```suggestion
		if !upstreamLatest || upstreamEnvName == "" {
```

Reply:
it’s ok if one of them is nil but not two of them, then we should throw an error

Reply:
You are totally right :D I was thinking of `!=` when I saw the unary operator. My bad.